internal/vault/auth_methods.go

```diff
@@ -0,0 +1,475 @@
+	}
+
+	for _, existingAuthMethod := range existingAuthList {
+		existingAuths[existingAuthMethod.Type] = existingAuthMethod
+	}
+
+	return existingAuths, nil
+}
+
+// getUnmanagedAuthMethods gets unmanaged auth methods by comparing what's already in Vault and what's in the externalConfig
+func (v *vault) getUnmanagedAuthMethods(managedAuthMethods []auth) map[string]*api.MountOutput {
+	// TODO: remove this line and just use passed value
+	unmangedAuths, _ := v.getExistingAuthMethods()
+	// Remove managed auth methods form the items since the rest will be disabled.
+	for _, mangedAuthMethod := range managedAuthMethods {
+		delete(unmangedAuths, mangedAuthMethod.Type)
+	}
+	// Remove token auth method since it's the default
+	delete(unmangedAuths, "token")
+
+	return unmangedAuths
+}
+
+func (v *vault) configureAuthMethods() error {
+	managedAuths := extConfig.Auth
+	existingAuths, _ := v.getExistingAuthMethods()
+	unManagedAuths := v.getUnmanagedAuthMethods(managedAuths)
+
+	err := v.syncManagedAuthMethods(managedAuths, existingAuths)
+	if err != nil {
+		return errors.Wrap(err, "error configuring managed auth methods")
+	}
+
+	err = v.disableUnmanagedAuthMethods(unManagedAuths)
+	if err != nil {
+		return errors.Wrap(err, "error while disabling unmanaged auth methods")
+	}
+
+	return nil
+}
+
+// Disables any auth method that's not managed if purgeUnmanagedConfig option is enabled
+func (v *vault) disableUnmanagedAuthMethods(unManagedAuths map[string]*api.MountOutput) error {
+	if len(unManagedAuths) == 0 || !extConfig.PurgeUnmanagedConfig.Enabled || extConfig.PurgeUnmanagedConfig.Exclude.Auths {
+		return nil
+	}
+
+	for authMethod := range unManagedAuths {
+		err := v.cl.Sys().DisableAuth(authMethod)
+		if err != nil {
+			return errors.Wrapf(err, "error disabling %s auth method in vault", authMethod)
+		}
+	}
+	return nil
+}
+
+func (v *vault) syncManagedAuthMethods(managedAuths []auth, existingAuths map[string]*api.MountOutput) error {
+	for _, authMethod := range managedAuths {
+		if len(authMethod.Path) == 0 {
+			authMethod.Path = authMethod.Type
+		}
+
+		description := fmt.Sprintf("%s backend", authMethod.Type)
+
+		// get auth mount options
+		// https://www.vaultproject.io/api/system/auth.html#config
+		var authConfigInput api.AuthConfigInput
+		hasMountOptions := authMethod.Options != nil
+		// https://www.vaultproject.io/api/system/auth.html
+		var options api.EnableAuthOptions
+		if hasMountOptions {
+			err := mapstructure.Decode(authMethod.Options, &authConfigInput)
+			if err != nil {
+				return errors.Wrap(err, "error parsing auth method options")
+			}
+			options = api.EnableAuthOptions{
+				Type:        authMethod.Type,
+				Description: description,
+				Config:      authConfigInput,
+			}
+		} else {
+			options = api.EnableAuthOptions{
+				Type:        authMethod.Type,
+				Description: description,
+			}
+		}
+
+		if existingAuths[authMethod.Type] == nil {
```

Comment:
What is the idea here? why do you need to sync existing auths?

Reply:
Since this loop and func is concerned only with managed auths, I have to make sure that the "EnableAuthWithOptions" won't run on any existing auth, as this will throw an error.

Reply:
clear :+1: 

then I'd suggest adding a comment here why that's needed (because if someone don't know it throws errors on existing mount it would be confusing)

Reply:
yep, good point, a comment is added, thanks for this ❤️